hwy/ops/ppc_vsx-inl.h

```diff
@@ -0,0 +1,4168 @@
+
+template <typename T, size_t N>
+HWY_API Mask128<T, N> Xor(Mask128<T, N> a, Mask128<T, N> b) {
+  return Mask128<T, N>{vec_xor(a.raw, b.raw)};
+}
+
+template <typename T, size_t N>
+HWY_API Mask128<T, N> ExclusiveNeither(Mask128<T, N> a, Mask128<T, N> b) {
+  return Mask128<T, N>{vec_nor(a.raw, b.raw)};
+}
+
+// ------------------------------ BroadcastSignBit
+
+template <size_t N>
+HWY_API Vec128<int8_t, N> BroadcastSignBit(Vec128<int8_t, N> v) {
+  return Vec128<int8_t, N>{
+    vec_sra(v.raw, vec_splats(static_cast<unsigned char>(7)))};
+}
+
+template <size_t N>
+HWY_API Vec128<int16_t, N> BroadcastSignBit(Vec128<int16_t, N> v) {
+  return Vec128<int16_t, N>{
+    vec_sra(v.raw, vec_splats(static_cast<unsigned short>(15)))};
+}
+
+template <size_t N>
+HWY_API Vec128<int32_t, N> BroadcastSignBit(Vec128<int32_t, N> v) {
+  return Vec128<int32_t, N>{vec_sra(v.raw, vec_splats(31u))};
+}
+
+template <size_t N>
+HWY_API Vec128<int64_t, N> BroadcastSignBit(Vec128<int64_t, N> v) {
+  return Vec128<int64_t, N>{vec_sra(v.raw, vec_splats(63ULL))};
+}
+
+// ------------------------------ ShiftLeftSame
+
+template <typename T, size_t N,
+          hwy::EnableIf<!hwy::IsFloat<T>() &&
+                        !hwy::IsSpecialFloat<T>()>* = nullptr>
+HWY_API Vec128<T, N> ShiftLeftSame(Vec128<T, N> v,
+                                   const int bits) {
+  using UnsignedLaneT =
+    typename detail::Raw128<MakeUnsigned<T>>::RawVectLaneType;
+  return Vec128<T, N>{vec_sl(v.raw, vec_splats(
+    static_cast<UnsignedLaneT>(bits)))};
+}
+
+// ------------------------------ ShiftRightSame
+
+template <typename T, size_t N, HWY_IF_UNSIGNED(T)>
+HWY_API Vec128<T, N> ShiftRightSame(Vec128<T, N> v,
+                                    const int bits) {
+  using UnsignedLaneT =
+    typename detail::Raw128<MakeUnsigned<T>>::RawVectLaneType;
+  return Vec128<T, N>{vec_sr(v.raw, vec_splats(
+    static_cast<UnsignedLaneT>(bits)))};
+}
+
+template <typename T, size_t N,
+          hwy::EnableIf<hwy::IsSigned<T>() &&
+                        !hwy::IsFloat<T>() &&
+                        !hwy::IsSpecialFloat<T>()>* = nullptr>
+HWY_API Vec128<T, N> ShiftRightSame(Vec128<T, N> v,
+                                    const int bits) {
+  using UnsignedLaneT =
+    typename detail::Raw128<MakeUnsigned<T>>::RawVectLaneType;
+  return Vec128<T, N>{vec_sra(v.raw, vec_splats(
+    static_cast<UnsignedLaneT>(bits)))};
+}
+
+// ------------------------------ ShiftLeft
+
+template <int kBits, typename T, size_t N,
+          hwy::EnableIf<!hwy::IsFloat<T>() &&
+                        !hwy::IsSpecialFloat<T>()>* = nullptr>
+HWY_API Vec128<T, N> ShiftLeft(Vec128<T, N> v) {
+  static_assert(0 <= kBits && kBits < sizeof(T) * 8, "Invalid shift");
+  return ShiftLeftSame(v, kBits);
+}
+
+// ------------------------------ ShiftRight
+
+template <int kBits, typename T, size_t N,
+          hwy::EnableIf<!hwy::IsFloat<T>() &&
+                        !hwy::IsSpecialFloat<T>()>* = nullptr>
```

Comment:
> Hi @johnplatts, this is awesome work, thank you and tip of the hat, looks like a great implementation. Great to see you've also successfully implemented runtime dispatch.
> 
> Just a few small change suggestions before we merge :)
> 
> FYI we have #1138 ready to go, which adds two ops. Do you have a reference to recommend for POWER intrinsics so I could attempt to implement them? (Last I saw, the official reference PDF was huge and very slow to browse)

The Power Vector Intrinsic Programming Reference that describes the Altivec/VSX intrinsics can be found at  https://files.openpower.foundation/s/9nRDmJgfjM8MpR7.

The Altivec Technology Programming Environments Manual, which describes the Altivec instructions, can be found at https://www.nxp.com/docs/en/reference-manual/ALTIVECPEM.pdf.

The Power Instruction Set Architecture Version 3.1B specification which describes the VSX instructions can be found at https://files.openpower.foundation/s/dAYSdGzTfW4j2r2.

Reply:
> This is happening often, want to add a HWY_IF variant that expands to this?

Yes

The reason why I have the ```!hwy::IsFloat<T>() && !hwy::IsSpecialFloat<T>()``` in multiple places is to avoid unnecessary repetition as the Altivec/VSX intrinsic functions are already overloaded for the different Altivec/VSX vector types (and the overloading also exists in C as a compiler extension).

The ```!hwy::IsSpecialFloat<T>()``` is there as we do not want some of the functions enabled for the ```float16_t``` and ```bfloat16_t``` types.